datastore/http/http.go

```diff
@@ -0,0 +1,174 @@
+package http
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/teslamotors/fleet-telemetry/metrics"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+// Producer client to handle http
+type Producer struct {
+	namespace        string
+	metricsCollector metrics.MetricCollector
+	logger           *logrus.Logger
+	records          chan *telemetry.Record
+	address          string
+	httpClient       http.Client
+}
+
+// Metrics stores metrics reported from this package
+type Metrics struct {
+	produceCount adapter.Counter
+	byteTotal    adapter.Counter
+	errorCount   adapter.Counter
+	bufferSize   adapter.Gauge
+}
+
+// Config contains the data necessary to configure an http producer.
+type Config struct {
+	// WorkerCount is the number of http producer routines running.
+	// This number should be increased if `http_produce_buffer_size` is growing.
+	WorkerCount int `json:"worker_count"`
+
+	// Address is the address to produce requests to.
+	Address string `json:"address"`
+
+	// Timeout is the number of seconds to wait for a response. Defaults to 10.
+	Timeout int `json:"timeout"`
+}
+
+const (
+	statusCodePreSendErr = "NOT_SENT"
+	bufferSize           = 128
+)
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// NewProducer sets up an HTTP producer
+func NewProducer(config *Config, metricsCollector metrics.MetricCollector, namespace string, logger *logrus.Logger) (telemetry.Producer, error) {
+	registerMetricsOnce(metricsCollector)
+
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	requests := make(chan *telemetry.Record, bufferSize)
+
+	producer := &Producer{
+		namespace:        namespace,
+		metricsCollector: metricsCollector,
+		records:          requests,
+		address:          config.Address,
```

Comment:
Why do you say we should prevent localhost?

Reply:
will that not mean http producer can call fleet telemetry server itself?

Reply:
The producer could call fleet telemetry server but I believe that would be rejected due to invalid certificates. And it would be useless.

The use case I am picturing is a server running with these processes:
- fleet-telemetry: exposed on port 4443
- custom-web-server: internally on port 3000. This is what receives http events and does something with them.